dali/python/nvidia/dali/pipeline.py

```diff
@@ -448,20 +465,48 @@ def _prepare_graph(self, define_graph = None):
                         edges.append(e)
                 else:
                     edges.append(edge)
+        ops.reverse()
+        self._ops = ops
+        self._graph_outputs = outputs
+        self._setup_input_callbacks()
+        self._py_graph_built = True
+
+    def _start_py_workers(self):
+        if not self._parallel_input_callbacks:
+            return
+        self._py_pool = WorkersPool.from_groups(
+            self._parallel_input_callbacks, self._prefetch_queue_depth, self._py_workers_init, self._py_workers_num)
+        # pool instance releases shared memory when garbage collected, thus it must outlive the pipeline instance
+        # when external source is used with no_copy=True
+        weakref.finalize(self, lambda pool : pool.close(), self._py_pool)
+        self._py_pool_started = True
+
+    def _init_pipeline_backend(self):
+        self._pipe = b.Pipeline(self._max_batch_size,
+                                self._num_threads,
+                                self._device_id,
+                                self._seed,
+                                self._exec_pipelined,
+                                self._prefetch_queue_depth,
+                                self._exec_async,
+                                self._bytes_per_sample,
+                                self._set_affinity,
+                                self._max_streams,
+                                self._default_cuda_stream_priority)
+        self._pipe.SetExecutionTypes(self._exec_pipelined, self._exec_separated, self._exec_async)
+        self._pipe.SetQueueSizes(self._cpu_queue_size, self._gpu_queue_size)
+        self._pipe.EnableExecutorMemoryStats(self._enable_memory_stats)
 
         # Add the ops to the graph and build the backend
         related_logical_id = {}
-        self._ops = []
-        while ops:
-            op = ops.pop()
-            self._ops.append(op)
+
+        for op in self._ops:
```

Comment:
Where did the topological sorting go?

Reply:
It's still in its place, reversing a list was added there.